### ISSUE TITLE
Suppress display of end-of-file newlines as blank lines

### ISSUE THREAD

**Issue by mainameiz** [atom/atom#7787]
Atom by default inserts a new line (a line break actually) at the end of file (and it's cool), but it also shows this newline as an empty line at the end of file, but it's prefereably don't do this. How can I fix this? Thank you very much.

**Comment by oBlissing** [comment]
Look here: #4741

**Comment by mainameiz** [comment]
It's not what I want. I **want a line break at the end of file** but I **don't want to show it as an empty line**

**Comment by mainameiz** [comment]
https://robots.thoughtbot.com/no-newline-at-end-of-file

```
So, it turns out that, according to POSIX, every text file (including Ruby and JavaScript source files) should end with a \n, or “newline” (not “a new line”) character. This acts as the eol, or the “end of line” character. It is a line “terminator”.
```

**Comment by biffen** [comment]
> There is a difference between *line break* and *empty line*

How do you mean? Could you show the hexdumps of a file that ends in a newline and one that ends with an empty line (encoded in, say, ASCII, and containing, say, `foo`). In my mind they're the same:

    00000000  66 6f 6f 0a                                       |foo.|
    00000004

**Comment by mainameiz** [comment]
@biffen Its not about how it represented in hex-dumps (its the same), Its about how an editor displays a *newline*, It should not show *empty line* because there is "\n" character at hte end of file

**Comment by biffen** [comment]
@mainameiz Then what is the difference between an empty line and a newline that isn't followed by anything?

**Comment by mnquintana** [comment]
I can see why you might want to hide it, but then how would you know that any file indeed is ending with a `\n` character as it should? Sure, if you saved a file with Atom with the default settings, if you "knew" beforehand you could assume that the newline was being added on save, but new users wouldn't have the prior knowledge to make this assumption. Instead they'd assume Atom was making no changes on save until taking a look at their git diff, which would make this feature all the more confusing.

I'd be :+1: for a package that does this, but very :-1: about changing this in core.

**Comment by mainameiz** [comment]
@mnquintana It must be an option of course.

**Comment by mainameiz** [comment]
@biffen 

```sh
> hexdump -c test.txt                                                                                     (master⚡)
0000000   1  \n                                                        
0000002
```

![vim ](http://dl.dropbox.com/u/5169025/%D0%92%D1%8B%D0%B4%D0%B5%D0%BB%D0%B5%D0%BD%D0%B8%D0%B5_056.jpg)

```sh
> hexdump -c test.txt                                                                                     (master⚡)
0000000   1  \n  \n                                                    
0000003
```

![vim](http://dl.dropbox.com/u/5169025/%D0%92%D1%8B%D0%B4%D0%B5%D0%BB%D0%B5%D0%BD%D0%B8%D0%B5_057.jpg)

**Comment by mainameiz** [comment]
I have changed my opinion it must be default behaviour :-) It is unix agreement and I don't see any reasons why somebody might want not to insert new line at the end of file.

**Comment by izuzak** [comment]
@mainameiz Just to make sure I understand your suggestion, can you clarify how these two files would be displayed and what the difference would be:

```
$ hexdump -c test1.txt
0000000   c  \n                                                        
0000002

$ hexdump -c test2.txt
0000000   c                                                            
0000001
```

**Comment by mainameiz** [comment]
Its will look like this ![vim](ttps://camo.githubusercontent.com/0644e232b221cad40ece7d0beb46d6b910bb05e5/687474703a2f2f646c2e64726f70626f782e636f6d2f752f353136393032352f2544302539322544312538422544302542342544302542352544302542422544302542352544302542442544302542382544302542355f3035362e6a7067)

**Comment by exogen** [comment]
As @mainameiz suggests, there is some precedence for this. In Vim, `test\n` will show one line, and not show a numbered line 2. If you do type `test` and press Enter to create a second line, there are actually two newlines (`test\n\n`). In Python, `"test\n".splitlines()` will return `['test']`, not `['test', '']`, etc.

**Comment by triplesek** [comment]
+1

**Comment by dylanwinn** [comment]
I would like to add that I appreciate that atom makes sure that there is a newline at the end of the file by default, a lot of text editors (nano) miss this kind of basic stuff. However, that newline should definitely not be rendered as an additional black line. Literally — by definition, a newline signifies the end of a line, *not* the beginning. I understand that the naming in confusing but all other Unix utilities work like this. If you open up a file in vim or emacs or run it through wc -l it will report one less line than atom shows.

**Comment by mnquintana** [comment]
One idea for this - instead of displaying a blank line, do what GitHub does and display nothing. If there is no trailing newline, then [display this icon](https://octicons.github.com/icon/no-newline/) at the end of the last line, like GitHub does in diffs.

**Comment by glen-84** [comment]
This is the main reason why I've been resisting enabling the "ensure single trailing newline" setting – the fact that there is a line number but no line content really bothers me.

To make it crystal clear:

Currently:

```
1. Line 1\n
2. Line 2\n
3.
```

What it should be:

```
1. Line 1\n
2. Line 2\n

```

**Comment by triplesek** [comment]
"ensure single trailing newline" enabled or not, atom will still show a 4th empty line if your 3rd line ends with a newline. The setting is helpful in ensuring it is not forgotten

**Comment by glen-84** [comment]
@triplesek I'm not sure what you're referring to. The setting is fine – the line numbering is not.

**Comment by triplesek** [comment]
@glen-84 What I'm saying is that this issue occurs whether or not you choose to enable the "ensure single trailing newline" setting. I don't see why this would cause you to resist enabling it, since presumably you'd want all of your lines to end with a `\n` anyway.

**Comment by glen-84** [comment]
@triplesek I couldn't get used to the idea of what looks like a blank line at the end of files, so I disabled that setting. However, I came to the realization that this is the correct behaviour, but the editor is incorrectly numbering the final line.

**Comment by spinningarrow** [comment]
I think GitHub does this the best - for files that _don't_ end in a \n, there's a little icon at the end of the last line. Files that do end with a \n don't show anything special (or any extra lines). If GitHub followed Atom's behaviour, all the files would show an extra line with a number in the end, which is confusing too.

**Comment by Pacoup** [comment]
Ignoring teletypewriter history, as far as I know, CP/M's filesystem could not record the size of a file, so you had to specify an end of file delimiter, normally Control-Z; i.e. 1A hex. This would give text files like this:

```
line[CR+LF]
line[CR+LF]
line[Control+Z]
```

But I believe this would have also been completely valid:

```
line[CR+LF]
line[CR+LF]
line[CR+LF][Control+Z]
```

Conversely, Unix filesystems did record file sizes so you didn't have to end files with an end of file delimiter. Instead, the end of record delimiter was used to end every line. Furthermore, Unix chose to omit the CR delimiter (purportedly, this would have been to save some memory). So text files looked like this:

```
line[LF]
line[LF]
line[LF]
```

Apple did the same but chose to omit LF instead, so files looked like this:

```
line[CR]
line[CR]
line[CR]
```

If you think about it, then, in CP/M, CR+LF meant **new line**, but in Unix, LF meant **end of line**. In this respect, it would have been correct for early display editors in Unix to display the above examples as three lines, but it would have also been correct for CP/M to display the above examples as four lines.

C, which became popular in the 70s/80s, was built for Unix, so it followed the mandatory delimiter practice, and I believe most if not all early compilers wouldn't work without the last line of code ending with an end of record delimiter. So you had CP/M derivatives which would show a new line after every end of record delimiter, and C, which would require an end of record delimiter for the last line of code, and every CP/M derivative developer would have gotten into the habit of putting an extra line at the end of their source code to please C compilers.

CP/M derivatives became incredibly popular thanks to IBM and Microsoft's success, so, that's a lot of developers putting new lines at the end of their code.

This seems to have been so pervasive that every single text editor and IDE except for Vim currently prints an extra line after LF. Yes, even Emacs, although it hides the number of the extra line in linum-mode. Smart.

So neither is right, historically-speaking, but on Unix, LF does mean end of a line. [The POSIX standard defines a line as](http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206):

> A sequence of zero or more non-&lt;newline> characters plus a terminating &lt;newline> character.

In other words, a sequence of zero or more non-&lt;newline> characters **without** a terminating &lt;newline> character is not a line, and text editors have no business showing a new line there on Unix-like systems.

But then there's the issue of what to do with CR+LF and Windows developers, and do you really want all those developers to start adding two LF at the end of every file because they think the new line is missing. Sigh…

So, maybe just an option to remove the extra visual line would be nice.

**Comment by exogen** [comment]
@Pacoup IMO it should match the conventions of the platform, possibly as you describe. In this case for UNIX it's easy: why not just match what `wc -l` and other tools show? Currently the answer is that they all show one less line than Atom does. Know of any equivalents on Windows that could be a basis for the behavior there?

**Comment by Pacoup** [comment]
Notepad? 😆

**Comment by Pacoup** [comment]
Right, so on a more serious note, to validate my theory, I booted up an MS-DOS 6.22 VM and this is what I got with the `edit` command, i.e. the MS-DOS Editor. This is interesting, because this thing used to be QBasic running in editor mode before 6.22.

First, by editing a simple file without creating any new line, you get this (when opened in Atom no less):

<img width="114" alt="screen shot 2016-09-22 at 23 19 00" src="https://cloud.githubusercontent.com/assets/112764/18773209/f7cadc00-811a-11e6-8ae6-13ab95fe2c75.png">

These two characters are CR+LF.

In fact, if we make a new file in MS-DOS Editor and try to move one line down, we get the same behaviour we see in all modern GUI editors and IDEs:

<img width="374" alt="1" src="https://cloud.githubusercontent.com/assets/112764/18773548/2be491b8-811e-11e6-8ea5-ccaebb18a0b6.png">

<img width="374" alt="2" src="https://cloud.githubusercontent.com/assets/112764/18773551/2f1c39c6-811e-11e6-9ced-4c765bfe981d.png">

The fact that it moves directly down is weird indeed, but the cursor can do so laterally as well, and new characters aren't created until you type, a peculiarity of QBasic I imagine. Nevertheless, it does not move further down than one line and this demonstrates the canonical behaviour we are looking for.

There is another interesting behaviour with MS-DOS Editor. If you open a file and type nothing, it won't let you move down one line. If you save it, it will have nothing in it, no CR+LF, which means MS-DOS Editor, and I presume QBasic in previous versions, follows the same behaviour, again, which is to display an extra line whenever there is a CR+LF sequence.

But this is different than Notepad. Notepad allows you to create delimiter-less files from its introduction in Windows 1.0 all the way up to Windows 10. Presumably, the difference could be attributed to QBasic's interpreter requiring an end of line character, as the MS-DOS Editor will forcibly add a CR+LF character when you add a new line.

In Windows 1.0, the cursor could not be moved down in a file with one line, and upon saving and examining the file with Atom, you can see that there is no CR+LF in sight.

<img width="145" alt="screen shot 2016-09-23 at 01 09 45" src="https://cloud.githubusercontent.com/assets/112764/18774801/beb22dd2-812a-11e6-92b1-89b859cd607e.png">

If we add a new line however, lo and behold, CR+LF comes back, which makes Notepad consistent with QBasic / MS-DOS Editor, save for the fact that Notepad is able to save a file devoid of delimiters, which perhaps not coincidentally is also a capability of virtually all text editors on Windows.

<img width="121" alt="screen shot 2016-09-23 at 01 18 30" src="https://cloud.githubusercontent.com/assets/112764/18774911/abe9579c-812b-11e6-922b-f0c1ce1be029.png">

Notepad behaves exactly the same way in Windows 10.

This is not completely representative of all CP/M / PC DOS / MS-DOS operating systems, however, as they all predate QBasic, but is sufficient to define Windows today. I presume QuickBASIC behaved the same way, and [the Wikipedia page on BASIC](https://en.wikipedia.org/wiki/BASIC) has some screenshots with the cursor on a new line, so perhaps this is entirely because of BASIC.

Therefore, we can conclude that the normal behaviour on Windows is to show a new line after CR+LF. That leaves LF up to interpretation, however, as LF-encoded files on Windows are usually of Unix provenance.

The platform would indicate that LF-encoded files should show LF as a new line, just like for CR+LF, but the source of LF-encoded files today is Unix-like systems, so to be technically accurate, Windows programs would have to display LF characters in LF-encoded files as line terminators, not new lines.

Conversely, this is not the case for Unix because under POSIX, a line ends with a &lt;newline> character, regardless of what the &lt;newline> format is. So in Unix, CR+LF must also be considered a line end, not a new line.

Unicode further defines other characters considered as &lt;newline> too, so it is definitely the case that under Unix, a file such as the delimiter-less text file outputted by Notepad is malformed.

Of course, in a perfect world, Microsoft would go POSIX-compliant and we could rename LF (Line Feed) to LE (Line End) and be done with this conversation forever. But, hey, #BASIC. I'm sure Dijkstra would laugh.

**Comment by JonasT** [comment]
Any chance this is fixed any time soon, at least for files with Linux line breaks?

Every other linux editor I have tested (vim, even gedit) does this differently, so this is a confusing render difference (even if I think the POSIX behavior is rather silly and I'd prefer to properly see a newline in all editors - but rendering it differently than all other editors is pretty confusing).

**Comment by JonasT** [comment]
@izuzak any chance this can be looked into? I think this is a prerequisite for getting https://github.com/atom/whitespace/issues/144 fixed (and in general it would be nice if Atom behaved consistent with other POSIX editors on Linux too)

**Comment by maxbrunsfeld** [comment]
Hi everyone, sorry that this issue has been ignored.

I agree that the extra line number can be a bit confusing. The fundamental UX problem is that in Atom, it is possible to represent buffers *without* a trailing newline. It's not a built-in assumption of the editor. So we need some way of distinguishing a buffer containing `hello` (no trailing newline) from a buffer containing `hello\n` (with a trailing newline).

We could show a 'No-Newline' icon in the case where the buffer lacks a trailing newline, but I feel like this adds unnecessary visual noise.

I'm leaning towards just hiding the line number for the final empty line in the case where the file does end with a newline. The only way to check for a trailing newline would be to try to move your cursor to the end of file, but that seems ok to me. Thoughts?

**Comment by JonasT** [comment]
I think the most important thing is here consistency with other editors. Therefore, a file without `\n` at the end is the absolutely rare invalid special case. (such a file cannot be created with any Linux editor, you'd need to hex dump edit it)

Therefore I think a 'No-Newline' icon is the superior solution since it should rarely happen in practice and it's worth highlighting it in the rare moments it is actually encountered.

**Comment by glen-84** [comment]
@JonasT You're assuming that everyone uses Linux. Having no EOF newline is not super unique IMO.

Having said that, a no-newline icon could certainly be an optional setting.

**Comment by JonasT** [comment]
@glen-84 I am suggesting this purely in the presence of Linux line breaks. So yes, I'm assuming Linux

**Comment by glen-84** [comment]
> macOS and Windows use different byte characters as line breaks anyway, don't they?

MacOS uses `\n` (same as Linux ... they used to use `\r`). Windows uses `\r\n`.

**Comment by JonasT** [comment]
Will `\n` at the end of a file show up as a visible new line on Mac OS X? Might be worth testing with hexdump (to check the actual `\n` characters in the binary contents).

I hope it behaves similar to Linux, but if it doesn't this might make this issue a lot more complicated.

**Comment by JonasT** [comment]
To sum it up:

1. Linux: single trailing `\n` should not lead to a visible trailing empty line
2. Windows: single trailing `\r\n` should lead to a visible trailing empty line
3. macOS: ??

**Comment by glen-84** [comment]
> Will \n at the end of a file show up as a visible new line on Mac OS X?

I'm almost certain that you'll see the same behaviour across all platforms.

**Comment by JonasT** [comment]
@glen-84 it certainly differs from Windows, where a single trailing `\r\n` actually leads to a visible trailing empty line (shown in Microsoft's own Notepad.exe):

![notepad](https://user-images.githubusercontent.com/1573352/30983136-d4a13d6c-a489-11e7-8f96-1b90a89e358a.png)

.. while a trailing `\n` on most older common Linux editors doesn't. So it doesn't seem to be fully platform-consistent..

**Comment by glen-84** [comment]
@JonasT See https://github.com/atom/whitespace/issues/144#issuecomment-332921140.

**Comment by maxbrunsfeld** [comment]
Yeah, AFAICT there's really nothing platform specific about this issue. This is purely a visual concern about line numbers in the gutter. We don't have a mandate to mimic any other particular editor on Windows, Linux or macOS. We just need to decide on an approach that makes sense from first principles. I am on board with changing our current rendering though, but I'm not really on board with doing it differently on different platforms because there is really no need for that.

**Comment by JonasT** [comment]
Yea I think Atom should follow the canonical `\r\n` interpretation of Microsoft Windows when running on Linux as well.

However, this leads us back to the issue that in this case, Windows line break files with active atom/whitespace package will be saved with what shows up in all Windows editors (and Atom on all platforms) as a trailing empty line, while Linux line break files with active atom/whitespace package would be saved with what is no visible trailing empty line in most Linux editors. This can be considered intended, but I think this visual inconsistency is bad.

**Comment by glen-84** [comment]
@maxbrunsfeld,

But if Windows doesn't follow POSIX (AFAIK), then Atom should show:

```
1. Line 1\r\n
2.
```

In Windows, and:

```
1. Line 1\n
```

On Linux/MacOSX.

... or am I missing something?

**Comment by JonasT** [comment]
@glen-84 yes that is exactly what I'm referring to, and why I made this ticket. (unless I'm misunderstanding what your comment referred to)

**Comment by maxbrunsfeld** [comment]
This issue is orthogonal to whether you're using `\r\n` or `\n` as your line ending.

Currently, if you open a file containing the text `hello\n`, we show this:

```
1| hello
2|
```

And if the file uses windows line endings (`hello\r\n`), we show the exact same thing. It doesn't matter at all what platform you're using.

The issue is that people find it confusing that we show the line number 2. I agree; I think we should simply hide the 2.

**Comment by glen-84** [comment]
@maxbrunsfeld,

It's not that simple.

1. Not showing the numbered line is the correct behaviour when using `\n` *only*, AFAIK. (see other Windows editors that will show the empty line [VS, etc.])
2. Just hiding the line number for `\n` could add more confusion, and lead to users (used to seeing a blank line) adding another and ending up with `\n\n`. There should perhaps be a visual indication that an EOF newline is present (a middle dot, different bg colour, etc.)

**Comment by maxbrunsfeld** [comment]
1. We don't need to do the same thing as windows editors just because you're editing a file with CRLF line endings. That just happens to be the UX of particular windows editor(s). We can have a different UX.

2. There is going to be *some* confusion no matter what we do. Atom is different from vim on purpose - trailing newlines are not a baked-in assumption, they are just the default behavior, provided via a package. There is no way for us to completely eliminate confusion for all users, but it seems like a fair number of people are getting confused by the fact that we should that final line number.

**Comment by JonasT** [comment]
I think what Vim etc should do for Windows line breaks derails us a bit. Nevertheless for what it's worth, I am quite sure most Windows editors behave like Notepad and Notepad is about the closest source of truth to get (since it's written by Microsoft after all). But I guess nobody disputes the current `\r\n` handling of Atom anyway.

**Comment by glen-84** [comment]
1. If it's different to almost every other Windows editor, that would quite likely confuse users.
2. Maybe some visual indication like this?

![atom-invis-line](https://user-images.githubusercontent.com/261509/30984812-c147482e-a48e-11e7-9c54-c8c289c6a866.png)

**Comment by JonasT** [comment]
What is the corresponding hex bytes for each example?

**Comment by maxbrunsfeld** [comment]
@JonasT As before, we're talking about a file with a single trailing newline (so `Line 1\n`, or `Line 1\r\n`, it doesn't matter).

I believe @glen-84 is showing 3 different possible renderings that we could adopt for this file.

**Comment by maxbrunsfeld** [comment]
Personally, I'm in favor of just leaving it blank, since the `•` symbol is already used for line continuations in the presence of soft wraps.

**Comment by JonasT** [comment]
Well I think none of those renderings should be used for `Line 1\r\n` since that should show an actual trailing empty line 2.

As for `Line 1\n`, I think the best choice is also none of those but simply showing only one line as other Linux editors.. (and an indicator for `Line 1` that indicates the missing`\n` when Atom runs on Linux or other `\n` characters are used previously for line breaks - similar to the GitHub web render UI)

**Comment by glen-84** [comment]
> What is the corresponding hex bytes for each example?

`\n` is decimal `10` or hex `a`
`\r` is decimal `13` or hex `d`

**Comment by perennialmind** [comment]
I am with @glen-84: `\r\n` signifies a line _separator_ and a bare `\n` is a line _terminator_. You basically have two distinct file formats for multiline text: the semantics are tied to the data, not the platform. Save a two-line file with Windows conventions, and you should get exactly one line break sequence (`1\r\n2`). Save the same file with Unix conventions, and you should get exactly two instances (`1\n2\n`).

The "ensure single trailing newline" is only relevant for `\r\n` encoded files, and should display as a third, blank line. The option is still valuable, but mostly for working with software of a Unix background like git.

Text files with bare `\n` breaks without a final `\n` are anomalous and may be malformed.

**Comment by JonasT** [comment]
@perennialmind that's pretty much what I keep suggesting, yes. The trailing `\n` missing is the special case and should be the only one that shows up with some sort of special marker, if at all. Otherwise, it should work exactly as you described.

> The "ensure single trailing newline" is only relevant for \r\n encoded files, and should display as a third, blank line. The option is still valuable, but mostly for working with software of a Unix background like git.

Not sure I agree with this one, because if it was implemented exactly like that it'd mean the end result shows up differently depending on whether you save the file with windows line breaks or unix line breaks.

*However*, I also think the behavior of the whitespace package should probably kept and discussed entirely separate and the more important thing is to fix the line rendering in the Atom core first, and wonder about the package later. (instead of catering the line rendering of Atom the other way around for whatever the whitespace package does).

**Comment by perennialmind** [comment]
@JonasT, I think I see where you're coming from. From the "file format" perspective, the save options should not affect the line/buffer display. Furthermore, having just discovered [.editorconfig](http://editorconfig.org), I see that there is another dimension of compatibility to worry about. Fun!

Would it be reasonable to make the line separator/terminator interpretation a feature of the core editor, normally determined based on detected line-break sequence? That should do the Right Thing™ 95% of the time and make the subtleties more explicit. Packages such as whitespace could provide a manual override for the remaining 5%. This might also make the implementation for the existing [editorconfig package](https://github.com/sindresorhus/atom-editorconfig) a bit cleaner.

As a user, I would expect all numbered lines to be "real". If Atom tells me a LF-terminated file has five lines, I expect Atom, `vim`, and `wc -l` to agree on that count. If Atom saves that same file with CRLF-separators, I expect Visual Studio to show me lines numbered `1` through `5`.

When saving a two-line file with the first line containing "**L1**" and the second **blank**:

|      | default  | as terminators | as separators |
| -----| -------- | ---------------| --------------|
| CRLF | `L1\r\n` | `L1\r\n\r\n`   | `L1\r\n`      |
| LF   | `L1\n\n` | `L1\n\n`       | `L1\n`        |

When reading a file, counted/numbered lines:

|                  | default | as terminators | as separators |
| ---------------- | ------- | -------------- | ------------- |
| `L1\r\n`         | 2       | 1              | 2             |
| `L1\r\n\r\n`     | 3       | 2              | 3             |
| `L1\r\nL2`       | 2       | 2\*            | 2             |
| `L1\n\n`         | 2       | 2              | 3             |
| `L1\n`           | 1       | 1              | 2             |
| `L1`             | 1       | 1\*            | 1             |
| `L1\nL2`         | 2\*     | 2\*            | 2             |
| `L1\nL2\r\n`     | 2 or 3  | 2              | 3             |

Cases with a "\*" imply a detectable deviation from expectations and, presumably, an applied correction. With the rules above, saving such files when otherwise unmodified, would change the contents on disk. That might merit a visual indicator of some kind. Either that or expose the terminator/separator option per-file, like vim does. One possibility would be to augment the line-ending indicator/selector widget with two more options for the special cases, makig it `CRLF`, `LF`, `CRLF-term`, and `LF-sep`.

I don't think that the mixed linebreak case has a correct heuristic, but on the flip side, there's no wrong answer either.

P.S. LF line separators might never make sense. You see CRLF terminators in the real world: email, http, and plenty of other RFCs call for end-of-line markers as CRLF. I maintain that the existing "ensure single trailing newline" option ought only apply to CRLF files, but as @JonasT says, that's a package concern.

P.P.S. Sorry for the insufferable detail. These are the details that drive me _nuts_.

**Comment by b1f6c1c4** [comment]
As a fundamentalist who thinks `\n` marks the termination of one single line, I wrote an atom package https://atom.io/packages/no-evil-eol-newline that gives a presentational fix for this issue. Hope that helps.

Everything here is about presentation. File content is not mutated by this package.

![image](https://user-images.githubusercontent.com/9152961/37565133-508f2d3c-2ade-11e8-8e3c-d7bbb9624270.png)

**Comment by JonasT** [comment]
@b1f6c1c4 yea, that's *exactly* how core Atom should behave (except maybe a nice icon instead of just red coloring would be nice for the missing `\n` case - but I understand your main priority was probably to get this working in the first place).

**Comment by lee-dohm** [comment]
Since @b1f6c1c4 was kind enough to write a package to handle this and since this is pretty low on the maintainer team's priority list, I'm going to close this for now. If anyone is interested in this functionality, please check out @b1f6c1c4's package [no-evil-eol-newline](https://atom.io/packages/no-evil-eol-newline) :+1:
